### ISSUE TITLE
is there any way to start the steps from a sns aws?

### ISSUE THREAD

**Issue by jeancarlosgarcia** [horike37/serverless-step-functions#106]
Are you planning to add to SNS events as serverless allows?

**Comment by horike37** [comment]
Would love to do that if CloudFormation supports SNS trigger for StepFunctions.

In order to achieve that, you need to specify something on StepFunctions to `Protocol` of `AWS::SNS::Subscription`. You can implement that if it is possible.
https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-sns-subscription.html#cfn-sns-protocol

**Comment by ksnyde** [comment]
I would love to see this too. Am I understanding the current limitation is that SNS doesn't support triggering a StepFunction yet?

![image](https://user-images.githubusercontent.com/980757/36922196-e51b9208-1e1b-11e8-99cc-f3dc977d3241.png)

I guess you could create a lambda function which just starts execution of the step function and then immediately executes. Seems sloppy though. 

Just out of interest, if a lambda function calls a StepFunction asynchronously can it exit immediately and the StepFunction completes execution beyond the execution lifetime of the lambda function?

**Comment by jeancarlosgarcia** [comment]
si claro puede funcionar perfectamente, y asi me toco hacer: tengo un lambda que se inicia apartir de un evento SNS y ese lambda invoca a la StepFunction, el lambda incicial muere y la StepFunction sigue su curso.

**Comment by jeancarlosgarcia** [comment]
Yes of course it can work perfectly, and so I have to do: I have a lambda that starts from an SNS event and that lambda invokes the StepFunction, the incidental lambda dies and the StepFunction takes its course.

**Comment by joshrussell** [comment]
This is possible now. I came across this limitation recently and for now I'll use the above workaround but CloudWatch Events now supports Step Functions as a target. Looks like as of 3/21.

**Comment by hasantayyar** [comment]
Yes better to create a design like this

`SNS Topic -> [event rule input] -> CloudWatch -> State machine`

**Comment by superhuman54** [comment]
@hasantayyar yes I heard CloudWatch supports State Machine as a target but all the examples are about the target. Can you explain CloudWatch Event Rule Pattern for SNS?

**Comment by hasantayyar** [comment]
@superhuman54 

![screen shot 2018-07-20 at 12 45 43](https://user-images.githubusercontent.com/88295/42998654-2976687a-8c1b-11e8-9bb4-09aae9bf3a80.png)

this should be possible via CloudFormation
But I believe you can not define the SNS topic here in the rule. so your state machine has to contain a logic and selectively continue or not based on context.

**Comment by ploh** [comment]
@hasantayyar @superhuman54 
Have you got "SNS publications as CloudWatch Event source" (and thereby as a trigger for Step Functions) working? I just tried two hours to come up with a rule event pattern for this but there do not seem to be any events that can be filtered on - even when enabling cloud trail.

**Comment by m4h3** [comment]
@ploh I think that doesn't work as SNS in cloudwatch only track admin action as events. https://docs.aws.amazon.com/sns/latest/dg/sns-logging-using-cloudtrail.html

**Comment by csabaszilveszter** [comment]
Any updates on this? Couldn't figure out if the `sns -> step functions` trigger is supported now or not just by looking at the documentation.

**Comment by zoadaare** [comment]
Echoing your question @csabaszilveszter . Maybe there is still an AWS engineer out there listening. There seems to be a missing piece between the sns and the cloudWatch rule. I created an sns topic without any subscriptions and a cloudWatch rule that has an event source as in @hasantayyar screenshot above. The event do not have the option to point to a specific topic and a subscription on the topic do not have cloudWatch as an option in the protocol list. I manually published a message, and the step function was not invoked. @AWSengineers, this is my design, s3_bucket (put event) -> sns-topic ->??? cloudWatch Event -> step function. I know I can fire the event directly from s3 through cloudTrail and cloudWatch to invoke the step function but my supervisor wants all events to be published through sns. If you have a working example, please do point us to the part of the documentation. Thanks

**Comment by zoadaare** [comment]
Echoing your question @csabaszilveszter . Maybe there is still an AWS engineer out there listening. There seems to be a missing piece between the sns and the cloudWatch rule. I created an sns topic without any subscriptions and a cloudWatch rule that has an event source as in @hasantayyar screenshot above. The event do not have the option to point to a specific topic and a subscription on the topic do not have cloudWatch as an option in the protocol list. I manually published a message, and the step function was not invoked. I edited the topic and the step function was invoked. I created an s3 event on the bucket to publish put events to the topic, this too did not invoke the step function. @AWSengineers, this is my design, s3_bucket (put event) -> sns-topic ->??? cloudWatch Event -> step function. I know I can fire the event directly from s3 through cloudTrail and cloudWatch to invoke the step function but my supervisor wants all events to be published through sns. If you have a working example, please do point us to the part of the documentation. Thanks

**Comment by patricknazar** [comment]
What about `[SNS Topic Subscription]` => `[SQS Queue]` => `[Lambda (node js SDK: stepfunctions.startExecution(...) )]` ? Let me know what you guys think of this approach.

**Comment by piersf** [comment]
This doesn't seem to be possible. Going to SNS Console and creating a subscription for a topic, it does not show Step Functions as an option in the drop-down menu.

It would be a great addition to have this implemented: SNS --> Step Function

**Comment by mbrice1024** [comment]
One thing I'd like to point out limitations with the proposed work around using CloudWatch:

- You can not ingest the payload of the published event. 
- You can't limit to just a publish API calls. So events like new subscription will trigger the pipeline.

**Comment by marcelinhov2** [comment]
Any updates?

**Comment by lee-11** [comment]
I understand that EventBridge is probably the "right" answer, but I would very much like to switch from Lambda to Step Functions in some older systems driven via SNS just to get out of the runtime and package updating. It's a path to simplify operations, whereas switching from SNS to EventBridge is a heavier lift with no such payoff.  So please do add the StepFunctions subscription type to SNS -- it would be helpful.

**Comment by VirtueMe** [comment]
SNS does not natively support Step Functions as a subscription target — SNS can invoke Lambda functions, SQS queues, HTTP endpoints, and similar, but not Step Functions directly. The typical pattern is to have SNS trigger a Lambda function which then starts the state machine via `StartExecution`. Closing as an AWS constraint outside the scope of this plugin.
